mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1167,6 +1197,223 @@ pub async fn p2sh_spending_tx<T: UtxoCommonOps>(coin: &T, input: P2SHSpendingTxI
+                return MmError::err(TxGenError::MinerFeeExceedsPremium {
+                    miner_fee: big_decimal_from_sat_unsigned(miner_fee, coin.as_ref().decimals),
+                    premium: args.premium_amount.clone(),
+                });
+            }
+            premium_sat - miner_fee
+        },
+    };
+
+    let dex_fee_address = address_from_raw_pubkey(
+        args.dex_fee_pub,
+        coin.as_ref().conf.pub_addr_prefix,
+        coin.as_ref().conf.pub_t_addr_prefix,
+        coin.as_ref().conf.checksum_type,
+        coin.as_ref().conf.bech32_hrp.clone(),
+        coin.addr_format().clone(),
+    )
+    .map_to_mm(|e| TxGenError::AddressDerivation(format!("Failed to derive dex_fee_address: {}", e)))?;
+    let dex_fee_output = TransactionOutput {
+        value: dex_fee_sat,
+        script_pubkey: Builder::build_p2pkh(&dex_fee_address.hash).to_bytes(),
+    };
+
+    let premium_address = address_from_raw_pubkey(
+        args.maker_pub,
+        coin.as_ref().conf.pub_addr_prefix,
+        coin.as_ref().conf.pub_t_addr_prefix,
+        coin.as_ref().conf.checksum_type,
+        coin.as_ref().conf.bech32_hrp.clone(),
+        coin.addr_format().clone(),
+    )
+    .map_to_mm(|e| TxGenError::AddressDerivation(format!("Failed to derive premium_address: {}", e)))?;
+    let premium_output = TransactionOutput {
+        value: premium_sat,
+        script_pubkey: Builder::build_p2pkh(&premium_address.hash).to_bytes(),
+    };
+
+    p2sh_spending_tx_preimage(coin, &prev_tx, lock_time, SEQUENCE_FINAL, vec![
+        dex_fee_output,
+        premium_output,
+    ])
+    .await
+    .map_to_mm(TxGenError::Legacy)
+}
+
+pub async fn gen_and_sign_dex_fee_spend_preimage<T: UtxoCommonOps>(
+    coin: &T,
+    args: &GenDexFeeSpendArgs<'_>,
+    htlc_keypair: &KeyPair,
+) -> GenAndSignDexFeeSpendResult {
+    let maker_pub = Public::from_slice(args.maker_pub).map_to_mm(|e| TxGenError::InvalidPubkey(e.to_string()))?;
+    let taker_pub = Public::from_slice(args.taker_pub).map_to_mm(|e| TxGenError::InvalidPubkey(e.to_string()))?;
+
+    let preimage = gen_dex_fee_spend_preimage(
+        coin,
+        args,
+        LocktimeSetting::CalcByHtlcLocktime(args.time_lock),
+        CalcPremiumBy::DeductMinerFee,
+    )
+    .await?;
+
+    let redeem_script = swap_proto_v2_scripts::dex_fee_script(args.time_lock, args.secret_hash, &taker_pub, &maker_pub);
+    let signature = calc_and_sign_sighash(
+        &preimage,
+        DEFAULT_SWAP_VOUT,
+        &redeem_script,
+        htlc_keypair,
+        coin.as_ref().conf.signature_version,
+        coin.as_ref().conf.fork_id,
+    )?;
+    let preimage_tx: UtxoTx = preimage.into();
+    Ok(TxPreimageWithSig {
+        preimage: serialize(&preimage_tx).take(),
+        signature: signature.take(),
+    })
+}
+
+pub async fn validate_dex_fee_spend_preimage<T: UtxoCommonOps + SwapOps>(
+    coin: &T,
+    gen_args: &GenDexFeeSpendArgs<'_>,
+    preimage: &TxPreimageWithSig,
```

Comment:
In UTXO case, maker can get a different transaction time lock and resulting premium amount (depending on miner fee). So, I had to add `LocktimeSetting` and `CalcPremiumBy` enums, and use the exact values from taker's preimage on maker's side. For other coin protocols, only signature might be sufficient.
https://github.com/KomodoPlatform/komodo-defi-framework/blob/ebe7e4f36b166c7262ec5b7182a4f16fee22ef46/mm2src/coins/utxo/utxo_common.rs#L1324-L1330

Reply:
> For other coin protocols, only signature might be sufficient.

Thanks for the explanation. I guess miner fee/gas discrepancy can also make signature not enough for other coins since they are part of the signed transaction. Can a different SIGHASH be used to allow bob decide the miner fee they want to pay instead of verifying the fee that alice calculated?

Reply:
For ETH/ERC-20, transaction preimage signing won't be required. Taker will sign a payload that will specify shares and destinations to which the payment should be spent. Maker will add the signature as an argument to smart contract call. For other protocols, additional research is required.

> Can a different SIGHASH be used to allow bob decide the miner fee they want to pay instead of verifying the fee that alice calculated?

I will check it on next iterations, thanks for the idea!